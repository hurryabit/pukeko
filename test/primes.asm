g_declare_cafs gm$cons_0_0, gm$cons_1_0, gm$abort, dict$Eq$Int, dict$Ord$Int, dict$Ring$Int, dict$Monoid$List, dict$Foldable$List, dict$Monad$IO, input, psums, primes, main, dict$Monad$IO$ll1
g_declare_main main

g_globstart gm$cons_0_0, 0
g_updcons 0, 0, 1
g_return

g_globstart gm$cons_0_1, 1
g_updcons 0, 1, 1
g_return

g_globstart gm$cons_0_2, 2
g_updcons 0, 2, 1
g_return

g_globstart gm$cons_0_4, 4
g_updcons 0, 4, 1
g_return

g_globstart gm$cons_1_0, 0
g_updcons 1, 0, 1
g_return

g_globstart gm$cons_1_2, 2
g_updcons 1, 2, 1
g_return

g_globstart gm$abort, 0
g_abort

g_globstart gm$eq, 2
g_push 1
g_eval
g_push 1
g_eval
g_eqv
g_update 3
g_pop 2
g_return

g_globstart gm$lt, 2
g_push 1
g_eval
g_push 1
g_eval
g_les
g_update 3
g_pop 2
g_return

g_globstart gm$le, 2
g_push 1
g_eval
g_push 1
g_eval
g_leq
g_update 3
g_pop 2
g_return

g_globstart gm$ge, 2
g_push 1
g_eval
g_push 1
g_eval
g_geq
g_update 3
g_pop 2
g_return

g_globstart gm$gt, 2
g_push 1
g_eval
g_push 1
g_eval
g_gtr
g_update 3
g_pop 2
g_return

g_globstart gm$neg, 1
g_eval
g_neg
g_update 1
g_return

g_globstart gm$add, 2
g_push 1
g_eval
g_push 1
g_eval
g_add
g_update 3
g_pop 2
g_return

g_globstart gm$sub, 2
g_push 1
g_eval
g_push 1
g_eval
g_sub
g_update 3
g_pop 2
g_return

g_globstart gm$mul, 2
g_push 1
g_eval
g_push 1
g_eval
g_mul
g_update 3
g_pop 2
g_return

g_globstart gm$mod, 2
g_push 1
g_eval
g_push 1
g_eval
g_mod
g_update 3
g_pop 2
g_return

g_globstart gm$seq, 2
g_eval
g_pop 1
g_update 1
g_unwind

g_globstart gm$puti, 1
g_eval
g_print
g_updcons 0, 0, 1
g_return

g_globstart gm$geti, 1
g_pop 1
g_input
g_update 1
g_return

g_globstart dict$Eq$Int, 0
g_pushglobal gm$eq, 2
g_push 0
g_updcons 0, 1, 2
g_pop 1
g_return

g_globstart dict$Ord$Int, 0
g_pushglobal gm$ge, 2
g_pushglobal gm$gt, 2
g_pushglobal gm$le, 2
g_pushglobal gm$lt, 2
g_push 0
g_push 2
g_push 4
g_push 6
g_updcons 0, 4, 5
g_pop 4
g_return

g_globstart dict$Ring$Int, 0
g_pushglobal gm$neg, 1
g_pushglobal gm$add, 2
g_pushglobal gm$sub, 2
g_pushglobal gm$mul, 2
g_push 0
g_push 2
g_push 4
g_push 6
g_updcons 0, 4, 5
g_pop 4
g_return

g_globstart dict$Monoid$List, 0
g_pushglobal gm$cons_0_0, 0
g_pushglobal dict$Monoid$List$ll1, 2
g_push 0
g_push 2
g_updcons 0, 2, 3
g_pop 2
g_return

g_globstart dict$Foldable$List, 0
g_pushglobal dict$Foldable$List$ll1, 3
g_pushglobal dict$Foldable$List$ll2, 3
g_push 0
g_push 2
g_updcons 0, 2, 3
g_pop 2
g_return

g_globstart dict$Monad$IO, 0
g_pushglobal dict$Monad$IO$ll2, 1
g_pushglobal dict$Monad$IO$ll4, 2
g_push 0
g_push 2
g_updcons 0, 2, 3
g_pop 2
g_return

g_globstart input, 0
g_pushglobal gm$cons_0_0, 0
g_pushglobal gm$geti, 1
g_pushglobal io$ll2, 2
g_updap 2, 1
g_unwind

g_globstart psums, 0
g_alloc 1
g_push 0
g_pushglobal psums$ll1, 4
g_updap 1, 1
g_pushint 0
g_pushglobal dict$Ring$Int, 0
g_push 2
g_updap 2, 2
g_pop 1
g_unwind

g_globstart primes, 0
g_pushglobal gm$cons_0_0, 0
g_pushint 4
g_cons 1, 2
g_pushint 2
g_cons 1, 2
g_pushglobal repeat$ll1, 1
g_mkap 1
g_pushint 5
g_cons 1, 2
g_pushglobal psums, 0
g_mkap 1
g_pushglobal sieve$ll2, 1
g_mkap 1
g_pushint 3
g_cons 1, 2
g_pushint 2
g_updcons 1, 2, 1
g_return

g_globstart main, 0
g_pushglobal main$ll1, 1
g_pushglobal input, 0
g_pushglobal dict$Monad$IO, 0
g_pushglobal bind$ll1, 1
g_updap 3, 1
g_unwind

g_globstart eq$ll1, 1
g_push 0
g_eval
g_proj 0
g_update 2
g_pop 1
g_unwind

g_globstart neq$ll1, 3
g_push 2
g_push 2
g_push 2
g_pushglobal eq$ll1, 1
g_mkap 3
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 1
g_pushglobal gm$cons_1_0, 0
g_update 4
g_pop 3
g_unwind
g_label .1
g_pop 1
g_pushglobal gm$cons_0_0, 0
g_update 4
g_pop 3
g_unwind
g_label .2

g_globstart le$ll1, 1
g_push 0
g_eval
g_proj 2
g_update 2
g_pop 1
g_unwind

g_globstart append$ll1, 1
g_push 0
g_eval
g_proj 1
g_update 2
g_pop 1
g_unwind

g_globstart add$ll1, 1
g_push 0
g_eval
g_proj 1
g_update 2
g_pop 1
g_unwind

g_globstart sub$ll1, 1
g_push 0
g_eval
g_proj 2
g_update 2
g_pop 1
g_unwind

g_globstart foldr$ll1, 1
g_push 0
g_eval
g_proj 0
g_update 2
g_pop 1
g_unwind

g_globstart foldl$ll1, 1
g_push 0
g_eval
g_proj 1
g_update 2
g_pop 1
g_unwind

g_globstart dict$Monoid$List$ll1, 2
g_push 0
g_push 2
g_pushglobal gm$cons_1_2, 2
g_pushglobal dict$Foldable$List, 0
g_pushglobal foldr$ll1, 1
g_updap 4, 3
g_pop 2
g_unwind

g_globstart dict$Foldable$List$ll1, 3
g_push 2
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 2
g_update 2
g_pop 1
g_unwind
g_label .1
g_uncons 2
g_push 1
g_push 4
g_push 4
g_pushglobal dict$Foldable$List, 0
g_pushglobal foldr$ll1, 1
g_mkap 4
g_push 1
g_push 4
g_updap 2, 6
g_pop 5
g_unwind
g_label .2

g_globstart dict$Foldable$List$ll2, 3
g_push 2
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 2
g_update 2
g_pop 1
g_unwind
g_label .1
g_uncons 2
g_push 1
g_push 1
g_push 5
g_push 5
g_mkap 2
g_push 4
g_pushglobal dict$Foldable$List, 0
g_pushglobal foldl$ll1, 1
g_updap 4, 6
g_pop 5
g_unwind
g_label .2

g_globstart nth_exn$ll1, 2
g_push 0
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 1
g_pushglobal gm$abort, 0
g_update 3
g_pop 2
g_unwind
g_label .1
g_uncons 2
g_pushint 0
g_push 4
g_pushglobal dict$Ord$Int, 0
g_pushglobal le$ll1, 1
g_mkap 3
g_eval
g_jumpcase .3, .4
g_label .3
g_pop 1
g_pushint 1
g_push 4
g_pushglobal dict$Ring$Int, 0
g_pushglobal sub$ll1, 1
g_mkap 3
g_push 2
g_pushglobal nth_exn$ll1, 2
g_updap 2, 5
g_pop 4
g_unwind
g_label .4
g_pop 1
g_update 4
g_pop 3
g_unwind
g_label .5
g_jump .2
g_label .2

g_globstart bind$ll1, 1
g_push 0
g_eval
g_proj 1
g_update 2
g_pop 1
g_unwind

g_globstart dict$Monad$IO$ll1, 0
g_pushglobal gm$cons_0_2, 2
g_update 1
g_unwind

g_globstart dict$Monad$IO$ll2, 1
g_push 0
g_pushglobal dict$Monad$IO$ll1, 0
g_updap 1, 2
g_pop 1
g_unwind

g_globstart dict$Monad$IO$ll3, 3
g_push 2
g_push 1
g_mkap 1
g_eval
g_uncons 2
g_push 1
g_push 1
g_push 5
g_updap 2, 6
g_pop 5
g_unwind

g_globstart dict$Monad$IO$ll4, 2
g_push 1
g_push 1
g_pushglobal dict$Monad$IO$ll3, 3
g_updap 2, 3
g_pop 2
g_unwind

g_globstart io$ll1, 3
g_push 1
g_push 1
g_mkap 1
g_push 3
g_push 1
g_cons 0, 2
g_push 1
g_pushglobal gm$seq, 2
g_updap 2, 5
g_pop 4
g_unwind

g_globstart io$ll2, 2
g_push 1
g_push 1
g_pushglobal io$ll1, 3
g_updap 2, 3
g_pop 2
g_unwind

g_globstart print$ll1, 1
g_push 0
g_pushglobal gm$puti, 1
g_pushglobal io$ll2, 2
g_updap 2, 2
g_pop 1
g_unwind

g_globstart repeat$ll1, 1
g_alloc 1
g_push 0
g_push 2
g_pushglobal dict$Monoid$List, 0
g_pushglobal append$ll1, 1
g_updap 3, 1
g_update 2
g_pop 1
g_unwind

g_globstart psums$ll1, 4
g_push 3
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 1
g_pushglobal gm$cons_0_0, 0
g_update 5
g_pop 4
g_unwind
g_label .1
g_uncons 2
g_push 4
g_push 1
g_push 5
g_pushglobal add$ll1, 1
g_mkap 3
g_push 2
g_push 1
g_push 6
g_push 6
g_mkap 3
g_push 1
g_updcons 1, 2, 8
g_pop 7
g_return
g_jump .2
g_label .2

g_globstart filter$ll1, 3
g_push 2
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 1
g_pushglobal gm$cons_0_0, 0
g_update 4
g_pop 3
g_unwind
g_label .1
g_uncons 2
g_push 1
g_push 4
g_mkap 1
g_push 1
g_push 4
g_mkap 1
g_eval
g_jumpcase .3, .4
g_label .3
g_pop 1
g_update 6
g_pop 5
g_unwind
g_label .4
g_pop 1
g_push 0
g_push 2
g_updcons 1, 2, 7
g_pop 6
g_return
g_jump .5
g_label .5
g_jump .2
g_label .2

g_globstart filter$ll2, 1
g_alloc 1
g_push 0
g_push 2
g_pushglobal filter$ll1, 3
g_updap 2, 1
g_update 2
g_pop 1
g_unwind

g_globstart sieve$ll1, 2
g_pushint 0
g_push 1
g_push 3
g_pushglobal gm$mod, 2
g_mkap 2
g_pushglobal dict$Eq$Int, 0
g_pushglobal neq$ll1, 3
g_updap 3, 3
g_pop 2
g_unwind

g_globstart sieve$ll2, 1
g_push 0
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 1
g_pushglobal gm$abort, 0
g_update 2
g_pop 1
g_unwind
g_label .1
g_uncons 2
g_push 1
g_push 1
g_pushglobal sieve$ll1, 2
g_mkap 1
g_pushglobal filter$ll2, 1
g_mkap 2
g_pushglobal sieve$ll2, 1
g_mkap 1
g_push 1
g_updcons 1, 2, 4
g_pop 3
g_return
g_jump .2
g_label .2

g_globstart main$ll1, 1
g_push 0
g_pushglobal primes, 0
g_pushglobal nth_exn$ll1, 2
g_mkap 2
g_pushglobal print$ll1, 1
g_updap 1, 2
g_pop 1
g_unwind
