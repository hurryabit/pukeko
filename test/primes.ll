external abort : ∀a. a = "abort"
external (+) : Int -> Int -> Int = "add"
external (-) : Int -> Int -> Int = "sub"
external (%) : Int -> Int -> Int = "mod"
external (!=) : Int -> Int -> Bool = "ne"
external (<=) : Int -> Int -> Bool = "le"
let nth$ll1 : ∀a. List a -> Int -> a =
      fun @a ->
        fun (xs : List a) (n : Int) ->
          match xs with
          | Nil @a -> abort @a
          | Cons @a x xs ->
            match (<=) n 0 with
            | False -> nth @a xs ((-) n 1)
            | True -> x
let nth : ∀a. List a -> Int -> a = nth$ll1
let append$ll1 : ∀a. List a -> List a -> List a =
      fun @a ->
        fun (xs : List a) (ys : List a) ->
          match xs with
          | Nil @a -> ys
          | Cons @a x xs -> Cons @a x (append @a xs ys)
let append : ∀a. List a -> List a -> List a = append$ll1
external print : Int -> IO Unit = "print"
external input : IO Int = "input"
external (>>=) : ∀a b. IO a -> (a -> IO b) -> IO b = "bind"
let repeat$ll1 : ∀a. List a -> List a =
      fun @a ->
        fun (xs : List a) ->
          let rec ys : List a = append @a xs ys in
          ys
let repeat : ∀a. List a -> List a = repeat$ll1
let psums$ll1 : (Int -> List Int -> List Int) -> Int -> List Int -> List Int =
      fun (psums0 : Int -> List Int -> List Int) (n : Int) (xs : List Int) ->
        match xs with
        | Nil @Int -> Nil @Int
        | Cons @Int x xs ->
          let y : Int = (+) x n in
          Cons @Int y (psums0 y xs)
let psums : List Int -> List Int =
      let rec psums0 : Int -> List Int -> List Int = psums$ll1 psums0 in
      psums0 0
let filter$ll1 : ∀a. (a -> Bool) -> (List a -> List a) -> List a -> List a =
      fun @a ->
        fun (p : a -> Bool) (filter_p : List a -> List a) (xs : List a) ->
          match xs with
          | Nil @a -> Nil @a
          | Cons @a x xs ->
            let ys : List a = filter_p xs in
            match p x with
            | False -> ys
            | True -> Cons @a x ys
let filter$ll2 : ∀a. (a -> Bool) -> List a -> List a =
      fun @a ->
        fun (p : a -> Bool) ->
          let rec filter_p : List a -> List a = filter$ll1 @a p filter_p in
          filter_p
let filter : ∀a. (a -> Bool) -> List a -> List a = filter$ll2
let sieve$ll1 : Int -> Int -> Bool =
      fun (p : Int) (k : Int) -> (!=) ((%) k p) 0
let sieve$ll2 : List Int -> List Int =
      fun (ks : List Int) ->
        match ks with
        | Nil @Int -> abort @(List Int)
        | Cons @Int p ks ->
          Cons @Int p (sieve (filter @Int (sieve$ll1 p) ks))
let sieve : List Int -> List Int = sieve$ll2
let primes : List Int =
      Cons @Int 2 (Cons @Int 3 (sieve (psums (Cons @Int 5 (repeat @Int (Cons @Int 2 (Cons @Int 4 (Nil @Int))))))))
let main$ll1 : Int -> IO Unit =
      fun (n : Int) -> print (nth @Int primes n)
let main : IO Unit = (>>=) @Int @Unit input main$ll1
