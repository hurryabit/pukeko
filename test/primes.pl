external abort : ∀a. a = "abort"
data Unit =
       | Unit
data Bool =
       | False
       | True
data Pair a b =
       | Pair a b
data Option a =
       | None
       | Some a
data Choice a b =
       | First a
       | Second b
data Dict$Eq a =
       | Dict$Eq (a -> a -> Bool)
eq : ∀a. Dict$Eq a -> a -> a -> Bool =
  fun @a ->
    fun (dict : Dict$Eq a) ->
      match dict with
      | Dict$Eq @a eq -> eq
neq : ∀a. Dict$Eq a -> a -> a -> Bool =
  fun @a ->
    fun (dict$Eq$a : Dict$Eq a) (x : a) (y : a) ->
      match eq @a dict$Eq$a x y with
      | False -> True
      | True -> False
data Dict$Ord a =
       | Dict$Ord (a -> a -> Bool) (a -> a -> Bool) (a -> a -> Bool) (a -> a -> Bool)
le : ∀a. Dict$Ord a -> a -> a -> Bool =
  fun @a ->
    fun (dict : Dict$Ord a) ->
      match dict with
      | Dict$Ord @a _ _ le _ -> le
data Dict$Monoid m =
       | Dict$Monoid m (m -> m -> m)
append : ∀m. Dict$Monoid m -> m -> m -> m =
  fun @m ->
    fun (dict : Dict$Monoid m) ->
      match dict with
      | Dict$Monoid @m _ append -> append
data Dict$Ring a =
       | Dict$Ring (a -> a) (a -> a -> a) (a -> a -> a) (a -> a -> a)
add : ∀a. Dict$Ring a -> a -> a -> a =
  fun @a ->
    fun (dict : Dict$Ring a) ->
      match dict with
      | Dict$Ring @a _ add _ _ -> add
sub : ∀a. Dict$Ring a -> a -> a -> a =
  fun @a ->
    fun (dict : Dict$Ring a) ->
      match dict with
      | Dict$Ring @a _ _ sub _ -> sub
data Int
external eq_int : Int -> Int -> Bool = "eq"
dict$Eq$Int : Dict$Eq Int =
  let eq : Int -> Int -> Bool = eq_int in
  Dict$Eq @Int eq
external lt_int : Int -> Int -> Bool = "lt"
external le_int : Int -> Int -> Bool = "le"
external ge_int : Int -> Int -> Bool = "ge"
external gt_int : Int -> Int -> Bool = "gt"
dict$Ord$Int : Dict$Ord Int =
  let ge : Int -> Int -> Bool = ge_int
  and gt : Int -> Int -> Bool = gt_int
  and le : Int -> Int -> Bool = le_int
  and lt : Int -> Int -> Bool = lt_int
  in
  Dict$Ord @Int ge gt le lt
external neg_int : Int -> Int = "neg"
external add_int : Int -> Int -> Int = "add"
external sub_int : Int -> Int -> Int = "sub"
external mul_int : Int -> Int -> Int = "mul"
dict$Ring$Int : Dict$Ring Int =
  let neg : Int -> Int = neg_int
  and add : Int -> Int -> Int = add_int
  and sub : Int -> Int -> Int = sub_int
  and mul : Int -> Int -> Int = mul_int
  in
  Dict$Ring @Int neg add sub mul
external mod : Int -> Int -> Int = "mod"
data Char
data Dict$Foldable t =
       | Dict$Foldable (∀a b. (a -> b -> b) -> b -> t a -> b) (∀a b. (b -> a -> b) -> b -> t a -> b)
foldr : ∀t. Dict$Foldable t -> (∀a b. (a -> b -> b) -> b -> t a -> b) =
  fun @t ->
    fun (dict : Dict$Foldable t) ->
      match dict with
      | Dict$Foldable @t foldr _ -> foldr
foldl : ∀t. Dict$Foldable t -> (∀a b. (b -> a -> b) -> b -> t a -> b) =
  fun @t ->
    fun (dict : Dict$Foldable t) ->
      match dict with
      | Dict$Foldable @t _ foldl -> foldl
data Dict$Functor f =
       | Dict$Functor (∀a b. (a -> b) -> f a -> f b)
data List a =
       | Nil
       | Cons a (List a)
dict$Monoid$List$ll1 : ∀a. List a -> List a -> List a =
  fun @a ->
    fun (xs : List a) (ys : List a) ->
      foldr @List dict$Foldable$List @a @(List a) (Cons @a) ys xs
dict$Monoid$List : ∀a. Dict$Monoid (List a) =
  fun @a ->
    let empty : List a = Nil @a
    and append : List a -> List a -> List a = dict$Monoid$List$ll1 @a
    in
    Dict$Monoid @(List a) empty append
dict$Foldable$List$ll1 : ∀a b. (a -> b -> b) -> b -> List a -> b =
  fun @a @b ->
    fun (f : a -> b -> b) (y0 : b) (xs : List a) ->
      match xs with
      | Nil @a -> y0
      | Cons @a x xs ->
        f x (foldr @List dict$Foldable$List @a @b f y0 xs)
dict$Foldable$List$ll2 : ∀a b. (b -> a -> b) -> b -> List a -> b =
  fun @a @b ->
    fun (f : b -> a -> b) (y0 : b) (xs : List a) ->
      match xs with
      | Nil @a -> y0
      | Cons @a x xs ->
        foldl @List dict$Foldable$List @a @b f (f y0 x) xs
dict$Foldable$List : Dict$Foldable List =
  let foldr : ∀a b. (a -> b -> b) -> b -> List a -> b =
        fun @a @b -> dict$Foldable$List$ll1 @a @b
  and foldl : ∀a b. (b -> a -> b) -> b -> List a -> b =
        fun @a @b -> dict$Foldable$List$ll2 @a @b
  in
  Dict$Foldable @List foldr foldl
nth_exn : ∀a. List a -> Int -> a =
  fun @a ->
    fun (xs : List a) (n : Int) ->
      match xs with
      | Nil @a -> abort @a
      | Cons @a x xs ->
        match le @Int dict$Ord$Int n 0 with
        | False -> nth_exn @a xs (sub @Int dict$Ring$Int n 1)
        | True -> x
data Dict$Monad m =
       | Dict$Monad (∀a. a -> m a) (∀a b. m a -> (a -> m b) -> m b)
bind : ∀m. Dict$Monad m -> (∀a b. m a -> (a -> m b) -> m b) =
  fun @m ->
    fun (dict : Dict$Monad m) ->
      match dict with
      | Dict$Monad @m _ bind -> bind
external seq : ∀a b. a -> b -> b = "seq"
external puti : Int -> Unit = "puti"
external geti : Unit -> Int = "geti"
data World =
       | World
data IO a = World -> Pair a World
dict$Monad$IO$ll1 : ∀a. a -> World -> Pair a World =
  fun @a -> Pair @a @World
dict$Monad$IO$ll2 : ∀a. a -> IO a =
  fun @a ->
    fun (x : a) ->
      coerce @(World -> Pair a World -> IO a) (dict$Monad$IO$ll1 @a x)
dict$Monad$IO$ll3 : ∀a b. IO a -> (a -> IO b) -> World -> Pair b World =
  fun @a @b ->
    fun (mx : IO a) (f : a -> IO b) (world0 : World) ->
      match coerce @(IO a -> World -> Pair a World) mx world0 with
      | Pair @a @World x world1 ->
        coerce @(IO b -> World -> Pair b World) (f x) world1
dict$Monad$IO$ll4 : ∀a b. IO a -> (a -> IO b) -> IO b =
  fun @a @b ->
    fun (mx : IO a) (f : a -> IO b) ->
      coerce @(World -> Pair b World -> IO b) (dict$Monad$IO$ll3 @a @b mx f)
dict$Monad$IO : Dict$Monad IO =
  let pure : ∀a. a -> IO a = fun @a -> dict$Monad$IO$ll2 @a
  and bind : ∀a b. IO a -> (a -> IO b) -> IO b =
        fun @a @b -> dict$Monad$IO$ll4 @a @b
  in
  Dict$Monad @IO pure bind
io$ll1 : ∀a b. (a -> b) -> a -> World -> Pair b World =
  fun @a @b ->
    fun (f : a -> b) (x : a) (world : World) ->
      let y : b = f x in
      seq @b @(Pair b World) y (Pair @b @World y world)
io : ∀a b. (a -> b) -> a -> IO b =
  fun @a @b ->
    fun (f : a -> b) (x : a) ->
      coerce @(World -> Pair b World -> IO b) (io$ll1 @a @b f x)
print : Int -> IO Unit = fun (n : Int) -> io @Int @Unit puti n
input : IO Int = io @Unit @Int geti Unit
repeat : ∀a. List a -> List a =
  fun @a ->
    fun (xs : List a) ->
      let rec ys : List a = append @(List a) (dict$Monoid$List @a) xs ys
      in
      ys
psums$ll1 : ∀_9. (∀_9. Dict$Ring _9 -> _9 -> List _9 -> List _9) -> Dict$Ring _9 -> _9 -> List _9 -> List _9 =
  fun @_9 ->
    fun (psums0 : ∀_9. Dict$Ring _9 -> _9 -> List _9 -> List _9) (dict$Ring$_9 : Dict$Ring _9) (n : _9) (xs : List _9) ->
      match xs with
      | Nil @_9 -> Nil @_9
      | Cons @_9 x xs ->
        let y : _9 = add @_9 dict$Ring$_9 x n in
        Cons @_9 y (psums0 @_9 dict$Ring$_9 y xs)
psums : List Int -> List Int =
  let rec psums0 : ∀_9. Dict$Ring _9 -> _9 -> List _9 -> List _9 =
            fun @_9 -> psums$ll1 @_9 psums0
  in
  psums0 @Int dict$Ring$Int 0
filter$ll1 : ∀a. (a -> Bool) -> (List a -> List a) -> List a -> List a =
  fun @a ->
    fun (p : a -> Bool) (filter_p : List a -> List a) (xs : List a) ->
      match xs with
      | Nil @a -> Nil @a
      | Cons @a x xs ->
        let ys : List a = filter_p xs in
        match p x with
        | False -> ys
        | True -> Cons @a x ys
filter : ∀a. (a -> Bool) -> List a -> List a =
  fun @a ->
    fun (p : a -> Bool) ->
      let rec filter_p : List a -> List a = filter$ll1 @a p filter_p in
      filter_p
sieve$ll1 : Int -> Int -> Bool =
  fun (p : Int) (k : Int) -> neq @Int dict$Eq$Int (mod k p) 0
sieve : List Int -> List Int =
  fun (ks : List Int) ->
    match ks with
    | Nil @Int -> abort @(List Int)
    | Cons @Int p ks ->
      Cons @Int p (sieve (filter @Int (sieve$ll1 p) ks))
primes : List Int =
  Cons @Int 2 (Cons @Int 3 (sieve (psums (Cons @Int 5 (repeat @Int (Cons @Int 2 (Cons @Int 4 (Nil @Int))))))))
main$ll1 : Int -> IO Unit =
  fun (n : Int) -> print (nth_exn @Int primes n)
main : IO Unit = bind @IO dict$Monad$IO @Int @Unit input main$ll1
