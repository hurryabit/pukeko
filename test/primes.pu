import std/prelude

repeat : List a -> List a
repeat xs =
  let rec ys = append xs ys in
  ys

psums : List Int -> List Int
psums =
  let rec psums0 n xs =
    match xs with
    | Nil -> Nil
    | Cons x xs ->
      let y = x + n in
      Cons y (psums0 y xs)
  in
  psums0 0

filter : (a -> Bool) -> List a -> List a
filter p =
  let rec filter_p xs =
    match xs with
    | Nil -> Nil
    | Cons x xs ->
      let ys = filter_p xs in
      if p x then Cons x ys else ys
  in
  filter_p

sieve : List Int -> List Int
sieve ks =
  match ks with
  | Nil -> abort
  | Cons p ks -> Cons p (sieve (filter (fun k -> k % p != 0) ks))

primes : List Int
primes = Cons 2 (Cons 3 (sieve (psums (Cons 5 (repeat (Cons 2 (Cons 4 Nil)))))))

main : IO Unit
main =
  input
  >>= fun n ->
  print (nth primes n)
