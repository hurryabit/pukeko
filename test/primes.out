59359

Steps  :   6918425398
Allocs :    108735539          (Checks:    108671970)
GC runs:            6
        2.28 real         2.01 user         0.26 sys
